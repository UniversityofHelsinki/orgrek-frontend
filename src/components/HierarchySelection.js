import * as React from 'react';
import Autocomplete from '@mui/material/Autocomplete';
import CheckBoxOutlineBlankIcon from '@mui/icons-material/CheckBoxOutlineBlank';
import CheckBoxIcon from '@mui/icons-material/CheckBox';
import TextField from '@mui/material/TextField';
import { connect, useDispatch } from 'react-redux';
import { useTranslation } from 'react-i18next';
import { Checkbox } from '@mui/material';

const icon = <CheckBoxOutlineBlankIcon fontSize="small" />;
const checkedIcon = <CheckBoxIcon fontSize="small" />;

const HierarchySelection = (props) => {
  const dispatch = useDispatch();
  const { t } = useTranslation();

  const iterate = (hierarchyList) => {
    let selectedHierarchies = [];
    for (const hierarchy of hierarchyList) {
      selectedHierarchies.push(hierarchy.value);
    }
    return selectedHierarchies.join(',');
  };

  if (props.selectableHierarchies?.length > 0) {
    const selectableHierarchiesList = props.selectableHierarchies
      .filter((item) => item !== 'history')
      .map((v) => ({
        value: v,
        label: t(v),
      }))
      .sort((a, b) =>
        t(a.label).toLowerCase() > t(b.label).toLowerCase() ? 1 : -1
      );

    let selectedHierarchies = [props.defaultHierarchy];
    if (props.selectedHierarchy) {
      selectedHierarchies = [...props.selectedHierarchy.split(',')];
    }

    const changeSelected = (event, hierarchies) => {
      const hierarchyList = hierarchies || props.defaultHierarchy;
      let selectedHierarchies;
      if (hierarchies && hierarchies.length > 0) {
        selectedHierarchies = iterate(hierarchyList);
      } else {
        selectedHierarchies = props.defaultHierarchy;
      }
      dispatch(dropDownSwitchValueCall(selectedHierarchies));
    };

    return (
      <Autocomplete
        multiple
        id="hierarchy-selection"
        disableCloseOnSelect
        options={selectableHierarchiesList}
        isOptionEqualToValue={(option, value) => option.value === value.value}
        getOptionLabel={(option) => option.label}
        value={selectedHierarchies.map((v) => ({ value: v, label: t(v) }))}
        onChange={changeSelected}
        renderOption={(props, option, { selected }) => (
          <li {...props}>
            <Checkbox
              icon={icon}
              checkedIcon={checkedIcon}
              style={{ marginRight: 8 }}
              checked={selected}
            />
            {option.label}
          </li>
        )}
        renderInput={(params) => (
          <TextField
            id="search"
            {...params}
            label={t('units')}
            placeholder=""
          />
        )}
      />
    );
  }
  return <></>;
};

export const dropDownSwitchValueCall = (data) => {
  return {
    type: 'SWITCH_HIERARCHY',
    payload: data,
  };
};

const mapStateToProps = (state) => ({
  selectedHierarchy: state.tree.selectedHierarchy,
  selectableHierarchies: state.tree.selectableHierarchies,
  defaultHierarchy: state.tree.defaultHierarchy,
});

export default connect(mapStateToProps)(HierarchySelection);
