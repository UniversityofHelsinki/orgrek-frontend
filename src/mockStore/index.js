export * from './mockApi';
export * from './mockDay';
export * from './mockEditMode';
export * from './mockHierarchy';
export * from './mockHierarchyFilters';
export * from './mockNode';
export * from './mockNodeDetailsView';
export * from './mockStore';
export * from './mockTree';
export * from './mockUser';
