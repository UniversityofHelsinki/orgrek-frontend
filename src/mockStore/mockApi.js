import { rest } from 'msw';
import format from 'date-fns/format';

const formatApiDate = (date) => format(date, 'd.M.yyyy');

const baseUrl = `${process.env.REACT_APP_ORGREK_BACKEND_SERVER || ''}/api`;

export const mockGetNode = (nodeId, body) =>
  rest.get(`${baseUrl}/node/${nodeId}`, (req, res, ctx) => res(ctx.json(body)));

export const mockGetAttributes = (nodeId, body, date) => {
  return rest.get(
    `${baseUrl}/node/${nodeId}/virallinen/${formatApiDate(
      date || new Date()
    )}/attributes`,
    (req, res, ctx) => res(ctx.json(body))
  );
};

export const mockPutNameAttributes = (nodeId) =>
  rest.put(`${baseUrl}/node/${nodeId}/attributes/names`, (req, res, ctx) =>
    res(ctx.delay(2000), ctx.json(req.json()))
  );

export const mockGetTypeAttributes = (nodeId, body) =>
  rest.get(`${baseUrl}/node/${nodeId}/attributes/types`, (req, res, ctx) =>
    res(ctx.json(body))
  );

export const mockPutTypeAttributes = (nodeId) =>
  rest.put(`${baseUrl}/node/${nodeId}/attributes/types`, (req, res, ctx) =>
    res(ctx.delay(2000), ctx.json(req.json()))
  );

export const mockGetValidHierarchyFilters = (date, body) =>
  rest.get(
    `${baseUrl}/hierarchyFilters/${formatApiDate(date)}`,
    (req, res, ctx) => res(ctx.json(body))
  );

export const mockGetCodeAttributes = (nodeId, body, date) =>
  rest.get(`${baseUrl}/node/${nodeId}/attributes/codes`, (req, res, ctx) =>
    res(ctx.json(body))
  );

export const mockSaveCodeAttributes = (nodeId) =>
  rest.put(`${baseUrl}/node/${nodeId}/attributes/codes`, (req, res, ctx) =>
    res(ctx.delay(2000), ctx.json(req.json()))
  );

export const mockGetOtherAttributes = (nodeId, selectedHierarchies, body) =>
  rest.get(
    `${baseUrl}/node/${nodeId}/attributes/others/hierarchies/${selectedHierarchies}`,
    (req, res, ctx) => res(ctx.json(body))
  );

export const mockSaveOtherAttributes = (nodeId) =>
  rest.put(`${baseUrl}/node/${nodeId}/attributes/others`, (req, res, ctx) =>
    res(ctx.delay(2000), ctx.json(req.json()))
  );

export const mockGetAttributeKeys = ({ selectedHierarchies, sections }, body) =>
  rest.get(
    `${baseUrl}/hierarchyFilters/${selectedHierarchies}/${
      (sections && sections.toString()) || 'virallinen'
    }/attributes/keys`,
    (req, res, ctx) => res(ctx.json(body))
  );

export const mockGetFullNames = (nodeId, date, body) =>
  rest.get(
    `${baseUrl}/node/fullname/${nodeId}/${formatApiDate(date)}`,
    (req, res, ctx) => res(ctx.json(body))
  );

export const mockGetAllFullNames = (nodeId, body) =>
  rest.get(`${baseUrl}/node/fullname/all/${nodeId}`, (req, res, ctx) =>
    res(ctx.json(body))
  );

export const mockGetFavorableFullNames = (nodeId, date, body) =>
  rest.get(
    `${baseUrl}/node/fullname/favorable/${nodeId}/${formatApiDate(date)}`,
    (req, res, ctx) => res(ctx.json(body))
  );

export const mockGetParents = (nodeId, date, hierarchies, body) =>
  rest.get(
    `${baseUrl}/node/${nodeId}/${date.toLocaleDateString(
      'fi-FI'
    )}/parents/${hierarchies}`,
    (req, res, ctx) => res(ctx.json(body))
  );

export const mockSaveParents = () =>
  rest.put(`${baseUrl}/node/parents/update`, (req, res, ctx) =>
    res(ctx.delay(2000), ctx.json(req.json()))
  );

export const mockGetChildren = (nodeId, date, hierarchies, body) =>
  rest.get(
    `${baseUrl}/node/${nodeId}/${date.toLocaleDateString(
      'fi-FI'
    )}/children/${hierarchies}`,
    (req, res, ctx) => res(ctx.json(body))
  );

export const mockSaveChildren = () =>
  rest.put(`${baseUrl}/node/children/update`, (req, res, ctx) =>
    res(ctx.delay(2000), ctx.json(req.json()))
  );

export const mockGetPredecessors = (nodeId, date, body) =>
  rest.get(
    `${baseUrl}/node/${nodeId}/${date.toLocaleDateString(
      'fi-FI'
    )}/predecessors`,
    (req, res, ctx) => res(ctx.json(body))
  );

export const mockGetSuccessors = (nodeId, date, body) =>
  rest.get(
    `${baseUrl}/node/${nodeId}/${date.toLocaleDateString('fi-FI')}/successors`,
    (req, res, ctx) => res(ctx.json(body))
  );

export const mockSaveSuccessors = (nodeId) =>
  rest.put(`${baseUrl}/node/${nodeId}/successors`, (req, res, ctx) =>
    res(ctx.delay(2000), ctx.json(req.json()))
  );

export const mockGetSectionTypeAttributes = (sectionType, body) =>
  rest.get(
    `${baseUrl}/node/section/${sectionType}/attributes`,
    (req, res, ctx) => res(ctx.json(body))
  );

export const mockGetTree = ({ hierarchies, selectedDay }, body) =>
  rest.get(
    `${baseUrl}/tree/${hierarchies}/${selectedDay.toLocaleDateString('FI-fi')}`,
    (req, res, ctx) => res(ctx.json(body))
  );

export const mockGetNodeValidity = (id, body) =>
  rest.get(`${baseUrl}/node/${id}`, (req, res, ctx) => res(ctx.json(body)));

export const mockSaveNodeValidity = (id) =>
  rest.put(`${baseUrl}/node/${id}/update`, (req, res, ctx) =>
    res(ctx.delay(2000), ctx.json(req.json()))
  );

/**
 * @deprecated not needed after everything has been migrated to RTK Query
 */
export const mockGetParentsDeprecated = (nodeId, date, hierarchies, body) =>
  rest.get(
    `${baseUrl}/node/parents/${nodeId}/${formatApiDate(date)}/${hierarchies}`,
    (req, res, ctx) => res(ctx.json(body))
  );
