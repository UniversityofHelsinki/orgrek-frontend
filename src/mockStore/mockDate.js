import React from 'react';

const RealDate = Date;

/**
 * Storybook decorator for mocking current time.
 *
 * Sets current time to a fixed value if systemTime is defined in story
 * parameters, otherwise uses the real time.
 */
export const withMockDate = (Story, { parameters }) => {
  if (parameters.systemTime instanceof Date) {
    // eslint-disable-next-line no-global-assign
    Date = class extends Date {
      constructor(...args) {
        if (args.length === 0) {
          super(parameters.systemTime.valueOf());
        } else {
          super(...args);
        }
      }

      static now() {
        return parameters.systemTime.getTime();
      }
    };
  } else {
    // eslint-disable-next-line no-global-assign
    Date = RealDate;
  }

  return <Story />;
};
