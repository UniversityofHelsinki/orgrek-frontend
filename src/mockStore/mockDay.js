import { createSlice } from '@reduxjs/toolkit';
import legacyReducer from '../reducers/dayChangeReducer';

const initialState = {
  selectedDay: new Date(),
};

const daySlice = createSlice({
  name: 'day',
  initialState,
  extraReducers(builder) {
    builder.addDefaultCase(legacyReducer);
  },
});

export const dayReducer = daySlice.reducer;
