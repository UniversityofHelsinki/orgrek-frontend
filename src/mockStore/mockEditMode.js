import { createSlice } from '@reduxjs/toolkit';
import legacyReducer from '../reducers/editModeReducer';

const initialState = {
  edit: false,
};

/**
 * @deprecated used only in the legacy edit view
 */
const editModeSlice = createSlice({
  name: 'editMode',
  initialState,
  extraReducers(builder) {
    builder.addDefaultCase(legacyReducer);
  },
});

/**
 * @deprecated used only in the legacy edit view
 */
export const editModeReducer = editModeSlice.reducer;
