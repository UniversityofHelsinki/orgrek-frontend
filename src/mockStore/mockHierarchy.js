const defaultHierarchies = [
  'tutkimus',
  'henkilosto',
  'toiminnanohjaus',
  'opetus',
  'history',
  'talous',
  'virallinen',
];

export const createHierarchies = (
  hierarchies = null,
  startDate = null,
  endDate = null
) =>
  (hierarchies || defaultHierarchies).map((hierarchy) => ({
    hierarchy,
    startDate,
    endDate,
  }));
