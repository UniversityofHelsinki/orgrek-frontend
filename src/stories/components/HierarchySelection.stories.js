import HierarchySelection from '../../components/HierarchySelection';
import { withMockStore } from '../../mockStore';

export default {
  component: HierarchySelection,
  argTypes: {
    size: {
      control: 'radio',
      options: ['small', 'medium'],
    },
  },
};

export const Default = {
  args: {
    size: 'medium',
  },
  decorators: [
    withMockStore({
      tree: {
        selectedHierarchy: 'virallinen',
        selectableHierarchies: [
          'tutkimus',
          'henkilosto',
          'toiminnanohjaus',
          'opetus',
          'history',
          'talous',
          'virallinen',
        ],
      },
    }),
  ],
};

export const Small = {
  ...Default,
  args: {
    ...Default.args,
    size: 'small',
  },
};
