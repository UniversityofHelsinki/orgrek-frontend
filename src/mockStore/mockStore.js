import { configureStore } from '@reduxjs/toolkit';
import { treeReducer } from './mockTree';
import { hierarchyReducer } from './mockHierarchy';
import { createNodeState, nodeReducer } from './mockNode';
import { userReducer } from './mockUser';
import { dayReducer } from './mockDay';
import { nodeDetailsViewReducer } from './mockNodeDetailsView';
import { hierarchyFiltersReducer } from './mockHierarchyFilters';
import { editModeReducer } from './mockEditMode';
import { api, notificationsReducer } from '../store';
import { Provider } from 'react-redux';
import React from 'react';
import { setupListeners } from '@reduxjs/toolkit/query';

export const configureMockStore = (preloadedState = {}) => {
  const store = configureStore({
    preloadedState,
    reducer: {
      [api.reducerPath]: api.reducer,
      tree: treeReducer,
      hr: hierarchyReducer,
      nrd: nodeReducer,
      ur: userReducer,
      dr: dayReducer,
      nvrd: nodeDetailsViewReducer,
      hierarchyFilters: hierarchyFiltersReducer,
      editModeReducer: editModeReducer,
      notifications: notificationsReducer,
    },
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware().concat(api.middleware),
  });

  setupListeners(store.dispatch);

  return store;
};

/**
 * Configures Redux store from initial state and any additional test data
 * passed to this component.
 *
 * Intended to be used as Storybook story decorator, but it should work in
 * tests as well.
 *
 * @param state
 * @param children a Storybook story or any component using the Redux store
 */
export const MockStoreProvider = ({ state, children }) => {
  const store = configureMockStore(state);

  return <Provider store={store}>{children}</Provider>;
};

/**
 * Storybook decorator function which provides the mock Redux store to the story
 *
 * @param state
 * @return a function that can be passed to Storybook decorator array
 */
export const withMockStore = (state) => {
  return function MockStoreDecorator(Story) {
    return (
      <MockStoreProvider state={state}>
        <Story />
      </MockStoreProvider>
    );
  };
};

/**
 * For convenience a decorator that mocks Redux store with the given user.
 *
 * Use this decorator when you need to mock only the current user.
 * Alternatively, withMockStore decorator can mock the whole state.
 *
 * @param user current user
 */
export const withUser = (user) => withMockStore({ ur: { user } });

/**
 * For convenience a decorator that mocks Redux store with the given node data.
 *
 * Use this decorator when you need to mock only the node reducer state.
 * Alternatively, withMockStore decorator can mock the whole state.
 *
 * @param nodeState overrides node reducer initial state
 */
export const withNode = (nodeState) =>
  withMockStore({ nrd: createNodeState(nodeState) });

/**
 * For convenience a decorator that mocks Redux store with the given hierarchy
 * data.
 *
 * Use this decorator when you need to mock only the hierarchy reducer state.
 * Alternatively, withMockStore decorator can mock the whole state.
 *
 * @param hierarchyState overrides hierarchy reducer initial state
 */
export const withHierarchy = (hierarchyState) =>
  withMockStore({ hr: { ...hierarchyState } });
