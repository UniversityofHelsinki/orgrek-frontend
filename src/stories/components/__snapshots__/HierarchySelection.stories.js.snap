// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stories/components/HierarchySelection HierarchySelection smoke-test 1`] = `
Object {
  "checked": undefined,
  "children": Array [
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "units",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "talous",
      "pressed": undefined,
      "role": "button",
      "value": undefined,
    },
    Object {
      "autocomplete": "list",
      "checked": undefined,
      "children": undefined,
      "haspopup": "listbox",
      "name": "units",
      "pressed": undefined,
      "role": "combobox",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "description": "Open",
      "name": "Open",
      "pressed": undefined,
      "role": "button",
      "value": undefined,
    },
  ],
  "name": "stories-components-hierarchyselection--hierarchy-selection",
  "pressed": undefined,
  "role": "WebArea",
  "value": undefined,
}
`;
